quil-rs/src/instruction/classical.rs

```diff
@@ -36,6 +36,21 @@ pub enum ArithmeticOperand {
     MemoryReference(MemoryReference),
 }
 
+impl std::hash::Hash for ArithmeticOperand {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match self {
+            Self::LiteralInteger(operand) => operand.hash(state),
+            Self::LiteralReal(operand) => hash_float(operand, state),
+            Self::MemoryReference(operand) => operand.hash(state),
+        }
+    }
+}
+
+#[inline]
+fn hash_float(value: &f64, state: &mut impl std::hash::Hasher) {
+    std::hash::Hash::hash(&format!("{value}"), state)
+}
+
```

Comment:
We can't derive `Hash` on structs containing a float because they don't have `Hash` implementations. This is in part because it's possible to write functions `f(a) != f(b)` for `a` and `b`, where `a == b`. This is mostly because of the semantics around `NaN`s and `+0` and `-0`. The quil spec isn't clear on what these values mean in the language, but I don't think they are generally cases we are too concerned with. Though, I could be wrong on that.

My thought, is that, for the purposes of `quil-rs`, if it formats two floats in the same way in a program, then it is effectively the same float. The downsides here are boilerplate and efficiency. However, in my survey of the issue, it seems that serialization  is among the top recommendations for handling this. 



Alternatively, there crates like [ordered-float](https://crates.io/crates/ordered-float) that wrap floats and offer implementations of `Hash`. However, I'm not sure if introducing another dependency, breaking changes, and the negative impact on ergonomics are worth it. 


For what it's worth, an experimental crate, [objecthash](https://crates.io/crates/objecthash) exists. It offers hash implementations for deeply nested structures, which can include float. From what I can understand, this also uses a fancy, cryptographically secure form of serialization.

Reply:
I think that the output of formatting being the hasher seems reasonable, since that covers the `+0` vs `-0` case (if a distinction is even necessary), and I'm not sure `NaN` is even allowed, just looking at the [grammar](https://github.com/quil-lang/quil/blob/289bc212017772c8e33f5d1a683f3dd4b8d83d42/grammars/Quil.g4#L230) it seems like no.

Reply:
I'm a fan of `ordered-float` for other reasons too, but I understand not wanting to pull in another dependency. For what it's worth, [here's](https://github.com/rigetti/quil-rs/blob/main/quil-rs/src/expression.rs#L155-L166) how we did it for `quil_rs::Expression::Number`; since the docs claim that `f64.to_bits()` is very portable, we just turned the floats to bits and hashed those.

Reply:
Ah, thanks for pointing me to that!. I forgot we had hashable expressions already. That looks promising, and I would rather have a consistent way of hashing floats across the board. I was worried about the portability of `to_bits`, despite what the docs say, but if it's been in use already, that calms the worries I had. I think what I'll do is lift that logic up so it can be shared with the instruction module.